Build against Y2038-compatible glibc for linux arm32

sbomer opened the pull request:
Reintroduces the fix for Y2038 support on arm32 linux (https://github.com/dotnet/runtime/pull/102059), which was reverted due to problems running against openssl built with _TIME_BITS=32. Thanks to @davidwrighton for help diagnosing the issue.

Fixes https://github.com/dotnet/runtime/issues/101444 (both the originally reported issue, and the test failures mentioned in https://github.com/dotnet/runtime/issues/101444#issuecomment-2111415106).

Supports: https://github.com/dotnet/runtime/issues/91826

sbomer committed:
Fix ARM ABI issue calling openssl function

sbomer committed:
Revert "Revert "Unpin arm build image, use Debian 12 helix images (#102059)" (#102324)"

This reverts commit 9a18cf5d88f1a43097646f147140a7078760ec4e.

jkotas commented on src/native/libs/System.Security.Cryptography.Native/openssl.c:
```diff
@@ -964,7 +964,21 @@ int32_t CryptoNative_X509StoreSetVerifyTime(X509_STORE* ctx,
+    // > If the argument requires double-word alignment (8-byte), the NCRN is rounded up to the next even register number.
+
+    // Let's pretend we're calling a function that takes both an int32_t and a time_t. This ensures that we set both
+    // r1 (to the low bits of our time_t) and r2:r3 (to the full time_t) which will satisfy either ABI.
```
Trimming the upper bits will produce a random bogus time. We may want to saturate the 32-bit value to max/min instead.

jkotas commented on src/native/libs/System.Security.Cryptography.Native/openssl.c:
```diff
@@ -964,7 +964,21 @@ int32_t CryptoNative_X509StoreSetVerifyTime(X509_STORE* ctx,
+
+    // Let's pretend we're calling a function that takes both an int32_t and a time_t. This ensures that we set both
+    // r1 (to the low bits of our time_t) and r2:r3 (to the full time_t) which will satisfy either ABI.
+    ((void (*)(X509_VERIFY_PARAM*, int32_t, time_t))(void*)(X509_VERIFY_PARAM_set_time))(verifyParams, (int32_t)verifyTime, verifyTime);
```
```suggestion
    // This can be deleted once the minimum supported Linux Arm32 distros are
    // at least Debian 13 and Ubuntu 24.04.
    ((void (*)(X509_VERIFY_PARAM*, int32_t, time_t))(void*)(X509_VERIFY_PARAM_set_time))(verifyParams, (int32_t)verifyTime, verifyTime);
```

sbomer committed:
Detect openssl time_t bitness

And avoid passing values that can't be represented in 32 bits
when running against openssl that expects 32-bit time_t.

sbomer committed:
Return failure code if for invalid time

sbomer committed:
Adjust comments

sbomer committed:
Fix whitespace

jkotas reviewed (approved):
LGTM. Thank you!

sbomer committed:
Move OPENSSL_gmtime out of FOR_ALL_OPENSSL_FUNCTIONS

Apparently this symbol did not exist in the builds of openssl
in Ubuntu 16.04.

sbomer committed:
Check FEATURE_DISTRO_AGNOSTIC_SSL

Except in opensslshim.c, which is only built for that feature.

jkotas commented on src/native/libs/System.Security.Cryptography.Native/opensslshim.c:
```diff
@@ -41,6 +44,15 @@ FOR_ALL_OPENSSL_FUNCTIONS
 #define MAKELIB(v)  SONAME_BASE v
 #endif
 
+#if defined(TARGET_ARM) && defined(TARGET_LINUX)
```
All these changes should be under `FEATURE_DISTRO_AGNOSTIC_SSL`.

When we are building against distro-specific OpenSSL, the `time_t` size is fixed. We do not need to be resilient variable `time_t` size.

jkotas replied:
Ah ok. Either way is fine then.

jkotas commented on src/native/libs/System.Security.Cryptography.Native/openssl.c:
```diff
@@ -7,10 +7,7 @@
 #include "pal_x509.h"
 #include "openssl.h"
-
-#ifdef FEATURE_DISTRO_AGNOSTIC_SSL
```
This looks like unintentional change. Revert?

jkotas commented:
We may want to simplify https://github.com/dotnet/runtime/blob/f02a695739d2e6c66de9deec14a7b0507c275a16/src/native/libs/System.Security.Cryptography.Native/pal_x509.c#L1357-L1364 to assume 64-bit `time_t` and instead add assert that will fail the build when `time_t` is not 64-bit. Something like this: https://github.com/jkotas/runtime/commit/8657f6d1aae71911fc85d460bd53b9a918e4207e

sbomer committed:
PR feedback

- Revert unintentional change

richlander commented:
Will you provided a detailed writeup on what we had to do? I think that will be super useful.
